docs/tutorials/gallery_adding_source_storages.py

```diff
@@ -0,0 +1,168 @@
+"""
+# Adding Source Storages
+
+Ragna has builtin support for [several source storage types][ragna.source_storages], but there may be cases
+where you want to use one that is not currently supported. 
+
+This tutorial walks you through the basics of adding a source storage that is not currently supported. 
+
+!!! note
+
+    This tutorial assumes that our `TutorialSourceStorage` class (shown below) is located in the file
+    `ragna/source_storages/_tutorial.py`.
+"""
+
+# %%
+# ## The Finished Product
+
+# %%
+# Here is the tutorial example printed in its entirety which will be explained below
+
+# %%
+# !!! note
+#
+#     The code snippet below only includes up to Step 1. To actually include your source storage
+#     in Ragna, you must still perform Step 2.
+
+from ragna.core import Source, SourceStorage
+
+
+class TutorialSourceStorage(SourceStorage):
+    def __init__(self):
+        # import database api
+
+        # set up database
+        ...
+
+    def store(self, sources: list[Source]) -> None:
```

Comment:
The `store()` and `retrieve()` methods of the `SourceStorage` ABC both have lists of `Document` types as parameters, but `Document` is part of `ragna.core._document`, which is a private module. I have instead replaced `Document` in the tutorial with `Source`, which is public. I did this (1) because I didn't want to make reference to any private parts of the codebase, and (2) because I wanted to be able to link to documentation of the types in the arguments to `store()` and `retrieve()`. @pmeier, do you have thoughts or input on this decision?

Reply:
> The `store()` and `retrieve()` methods of the `SourceStorage` ABC both have lists of `Document` types as parameters, but `Document` is part of `ragna.core._document`, which is a private module.

That is true, but they are exposed under `ragna.core`

https://github.com/Quansight/ragna/blob/5a9bdf69de6b4234831532c0d13797a7015a61bf/ragna/core/__init__.py#L34-L35

In general, we expose all the functionality that users need either under `ragna.core` or `ragna.deploy`. Anything below that is private. This has two upsides for us

1. We expose only exactly what we want given that we need put it in explicitly.
2. We can freely change module structure or the like in our implementation and only need to keep the upper namespace BC compliant.

> I have instead replaced `Document` in the tutorial with `Source`, which is public.

The two are two completely different objects and cannot be used interchangeably:
- A `Document` is the starting point that we extract information from. Right now this is handled by source storage directly, but that might change soon (see #191).
- A `Source` represents the information that we have extracted earlier.